### DISCUSSION TITLE
BoxShadow or drop shadow for vector icons

### DISCUSSION THREAD

**Discussion by flarive** [AvaloniaUI/Avalonia#11991] [Q&A]
Hello,

I'm trying to create a kind of drop shadow for vector icons for my neumorphic avalonia theme.
I'm using Avalonia 0.11 rc1.

Here is what i'm trying to do :

![3aeb44147160445](https://github.com/AvaloniaUI/Avalonia/assets/7724904/8ec01ae7-49b6-459d-9617-52c2840d6081)

![test](https://github.com/AvaloniaUI/Avalonia/assets/7724904/2b05f0b5-0aac-4108-988f-b0fa0d5b5efa)

How could i acheive that kind of effect please ?

A kind of BoxShadow working with Geometry, StreamGeometry, GeometryDrawing or Path would be so great !

Thanks for your help

Eviral

**Comment by dbriard** [comment]
I have a way to do that.
Basically, I duplicated the Shape classes of avalonia into SkiaShape classes, and I changed the Render to use a custom draw operation.
```
public sealed override void Render(DrawingContext context)
        {
            var geometry = RenderedGeometry;

            if (geometry != null)
            {
                var stroke = Stroke;

                ImmutablePen? pen = null;

                if (stroke != null)
                {
                    pen = new ImmutablePen(
                        stroke.ToImmutable(),
                        StrokeThickness);
                }
                //context.DrawGeometry(Fill, pen, geometry);
                context.Custom(new ShapeCustomDrawOperation(geometry, Bounds, Blur, Shadow)
                {
                    Pen = pen,
                    Brush = Fill,
                    Opacity = (float) this.Opacity,
                });
            }
        }
```

For the custom draw operation, you can found an example in avalonia's samples by looking for "ICustomDrawOperation" and you can also look for the code to draw the shape in avalonia source, in DrawingContextImpl.cs of Skia I think.

Once you can reproduce the rendering of the shape, you just have to add a DropShadow to the SKPaint object
     paint.ImageFilter = SKImageFilter.CreateDropShadow((float) shadow.OffsetX, (float) shadow.OffsetY, blur, blur, shadow.Color.ToSKColor(), paint.ImageFilter);
                      
                      Something like that in the custom draw op Render
                      
                     ```
 using (SKPaint paint = new SKPaint())
                    {
                        paint.Color = new SKColor(255, 255, 255, (byte) (255 * Opacity));

                        if (Shadow is BoxShadow shadow)
                        {
                            float blur = SkiaControl.BlurRadiusToSigma((float) shadow.Blur);
                            paint.ImageFilter = SKImageFilter.CreateDropShadow((float) shadow.OffsetX, (float) shadow.OffsetY, blur, blur, shadow.Color.ToSKColor(), paint.ImageFilter);
                        }

                        SKPicture picture;

                        using (var recorder = new SKPictureRecorder())
                        {
                            var cullRect = UnmodifiedBounds;

                            using (var recordingCanvas = recorder.BeginRecording(cullRect.ToSKRect()))
                            {
                                DrawShape(recordingCanvas);
                            }
                            picture = recorder.EndRecording();
                        }

                        canvas.DrawPicture(picture, paint);
                    }
```

It require a bit of work to prepare the classes but once you get custom draw elements working, you can do many cool stuffs.

**Reply by flarive** [reply]
Gonna try, thanks a lot !

**Reply by flarive** [reply]
How can i have a reference to Avalonia.Base ?
I can't resolve a lot of classes, i suppose they are defined in Avalonia.Base but it's scope is internal I think.

**Reply by dbriard** [reply]
Hum, I didn't have such issue. I should give a try to a new project to verify.

SkiaSharp 2.88.3 need to be installed.

The only non-public thing I needed was the SKPath of the Shape and I had to use reflection for that:

```
public static SKPath? GetStrokePath(this Geometry geometry)
        {
            s_geometryPlatformImpl ??= typeof(Geometry).GetProperty("PlatformImpl", BindingFlags.NonPublic | BindingFlags.Instance);
            Debug.Assert(s_geometryPlatformImpl is not null);
            var platformImpl = s_geometryPlatformImpl.GetValue(geometry);
            Debug.Assert(platformImpl is not null);
            var s_geometryPlatformImplStrokePath = platformImpl.GetType().GetProperty("StrokePath");
            Debug.Assert(s_geometryPlatformImplStrokePath is not null);
            return s_geometryPlatformImplStrokePath.GetValue(platformImpl) as SKPath;
        }
```

**Reply by dbriard** [reply]
There is this part where you need to install System.Reactive package.
Avalonia have a Avalonia.Reactive extensions that replace System.Reactive but they are not public unfortunatly... 

```
protected static void AffectsGeometry<TShape>(params AvaloniaProperty[] properties)
            where TShape : SkiaShape
        {
            foreach (var property in properties)
            {
                property.Changed.Subscribe(e =>
                {
                    if (e.Sender is TShape shape)
                    {
                        AffectsGeometryInvalidate(shape, e);
                    }
                });
            }
        }
```

**Comment by dbriard** [comment]
Here is a sample project inspired by the CustomDrawOp sample and Avalonia.Svg project on github: [AvaloniaBetterShapes.zip](https://github.com/AvaloniaUI/Avalonia/files/11970387/AvaloniaBetterShapes.zip)

The project only include Ellipse, but you can do the same for Path, Rectangle, etc... 
![image](https://github.com/AvaloniaUI/Avalonia/assets/3595443/350b16af-af0b-491e-9799-3aadcdbeb1bc)

The project only use the first BoxShadow in BoxShadows, but if you use all of them, don't hesitate to share the modification here ;)

**Reply by flarive** [reply]
Thank you so much !!!
